hathor/cli/top.py

```diff
@@ -0,0 +1,704 @@
+        """Send a cmd to the profiler and logs the results."""
+        assert Color is not None
+        now = datetime.datetime.now()
+        now_str = now.astimezone().strftime("%Y-%m-%d %H:%M:%S %Z")
+        self.log(now_str, Color.CMD_DATETIME)
+
+        cmd_fn = self.cmd_map.get(cmd, None)
+        self.log('> sending {} command...'.format(cmd), Color.CMD_CMDLINE)
+        if cmd_fn is None:
+            self.log('command not found: {}'.format(cmd))
+            return
+        try:
+            ret = await cmd_fn()
+            self.log(str(ret))
+        except Exception as e:
+            self.log('error: {}'.format(e))
+        finally:
+            self.log('')
+
+    def print_cmd(self, cmd: str, description: str) -> None:
+        """Util to print a cmd."""
+        assert Color is not None
+        self.win.addstr(cmd, Color.HELP_BOLD)
+        self.win.addstr(': ', Color.HELP_BOLD)
+        self.win.addstr(description)
+        self.win.addstr('\n')
+
+    def render(self) -> None:
+        assert Color is not None
+        self.win.addstr('profiler control\n', Color.HELP_BOLD)
+        self.win.addstr('\n')
+        self.print_cmd('c', 'clear log')
+        self.print_cmd('w', 'send a start command')
+        self.print_cmd('o', 'send a stop command')
+        self.print_cmd('r', 'send a reset command')
+        self.print_cmd('q', 'go back to main window')
+        self.win.addstr('\n')
+
+        height, width = self.win.getmaxyx()
+        y, _ = self.win.getyx()
+
+        self.win.addstr('─' * width)
+        txt = ' OUTPUT '
+        self.win.move(y, (width - len(txt)) // 2)
+        self.win.addstr(txt)
+        self.win.move(y + 1, 0)
+
+        y, _ = self.win.getyx()
+        max_rows = height - y - 1
+        for msg, attr in self._logs[-max_rows:]:
+            self.win.addstr(msg, attr)
+            self.win.addstr('\n')
+
+
+class ScreenManager:
+    """Manage the screen."""
+
+    def __init__(self, loop: AbstractEventLoop, fetcher: 'DataFetcher', *, update_interval: int = 2) -> None:
+        """Initialize the screen and immediately draw an initial screen."""
+        self.stdscr = curses.initscr()
+        curses.noecho()
+        curses.cbreak()
+        curses.curs_set(0)
+        self.stdscr.keypad(True)
+        self.stdscr.nodelay(True)
+        self.register_colors()
+
+        self._redraw: bool = False
+
+        self.freeze_data: bool = False
+
+        self.loop: AbstractEventLoop = loop
+        self.loop.add_reader(sys.stdin, self.getch)
+
+        self.fetcher: 'DataFetcher' = fetcher
+
+        self.screen_list: Dict[str, Window] = {
+            'help': HelpWindow(self, self.stdscr),
+            'main': MainWindow(self, self.stdscr),
+            'control': ControlWindow(self, self.stdscr),
+        }
+        self.screen = self.screen_list['main']
+
+        # Capture Ctrl+C.
+        def signal_handler(sig, frame):
+            self.quit()
+        signal.signal(signal.SIGINT, signal_handler)
```

Comment:
This is a bit odd and it very much looks broken: `signal_handler` has no access to `self` and I think it should be marked with `@staticmethod` (but I'm not sure what happens when accessing the bare function before the class can bind it). Either way it will also end-up creating a method that when called as `screen_manager.signal_handler(frame)` would have `sig` set to what `self` normally is, but I'd still expect it to fail because `self` wouldn't be defined. Maybe `signal.signal(...)` should be called inside `__init__` and add `self` arg to this function. Maybe it was like this before and a refactor broke it.

Reply:
Ohh, it's already inside `__init__`. NVM

Reply:
I guess you haven't noticed that it's already inside `__init__`. Maybe we should move it outside but it works the way it is.